feat: content gifting

miguelpeixe opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

NPPD-722

Allow subscribers to gift articles for 24 hours, up to a configured limit:

<img width="1080" height="779" alt="image" src="https://github.com/user-attachments/assets/28763a5c-1bb0-44ce-8f45-db2294cd0a17" />

Gifting an article creates a user meta entry with a key to unlock it. This key is valid for 24 hours and is shared via a query argument formatted as such: `{user_id}|{key}`, e.g.: `?content_key=79|uM5cAOZZy73AOiWGG13inWc9blf7RoS5`

## Gifting an article

### Jetpack sharing module

<img width="993" height="354" alt="image" src="https://github.com/user-attachments/assets/62b8e772-589d-4a39-8b2a-77465e197b79" />

#### Styles

| | |
| --- | --- |
| Icon + Text | <img width="337" height="62" alt="image" src="https://github.com/user-attachments/assets/dccfbc3d-474e-43a3-a9b9-a65dbbfd68cb" /> |
| Icon only | <img width="258" height="67" alt="image" src="https://github.com/user-attachments/assets/1dc76519-70b4-44d3-b2cb-ddadee022bec" /> |
| Text only | <img width="265" height="64" alt="image" src="https://github.com/user-attachments/assets/54560205-0891-40d9-83bd-973b1457a42d" /> |
| Official buttons | <img width="280" height="55" alt="image" src="https://github.com/user-attachments/assets/c87eca12-b5f1-4eeb-9d27-3fc01e6175e8" /> |

When enabling the gifting feature, the "Newspack Gift Article" sharing service is automatically added to the enabled sharing services. The sharing service is also added to the defaults, in case nothing is configured in the database.

### Without Jetpack sharing module

<img width="1068" height="429" alt="image" src="https://github.com/user-attachments/assets/4a9fad54-81b9-4dcd-8f99-d95abec571aa" />

## Gifting modal

<img width="1074" height="634" alt="image" src="https://github.com/user-attachments/assets/6b3b193d-039e-4752-9361-538458b92112" />

### Limit exceeded

<img width="443" height="378" alt="image" src="https://github.com/user-attachments/assets/fb231845-55e0-44e4-90ec-e93c0bfbb446" />

## Viewing a gifted article

When viewing a gifted article, you get a configurable CTA. If no URL is set, it will look for the primary subscription product to trigger a modal checkout.

### Light

<img width="1201" height="604" alt="image" src="https://github.com/user-attachments/assets/6cfac927-8668-4c0b-84ea-d76a60b9f155" />

### Dark

<img width="1194" height="618" alt="image" src="https://github.com/user-attachments/assets/4f2b729e-496a-4907-8cef-d9b6455ef5e3" />

### How to test the changes in this Pull Request:

### Default settings

1. Make sure you have:
   1. the Jetpack sharing module enabled
   2. metering disabled
   3. a primary subscription product (Audience -> Subscriptions)
2. Navigate to Audience -> Content Gate and enable Content Gifting
3. As a member, navigate to a free article (not locked behind a gate)
4. Confirm the button is not present, as this is not gated content
5. Navigate to a restricted article and confirm the button renders next to other sharing services
6. Click the button and confirm the modal renders with correct information and the "Copy link" button adds the string to the clipboard
7. Paste the link in an anonymous session
8. Confirm you get the full article with the default CTA. Clicking "Subscribe now" starts the modal checkout for the primary subscription product

### Custom Jetpack styles

1. Go to the Jetpack sharing module settings
2. Test all the different style options and confirm they render as expected in the subscriber session

### Without Jetpack sharing

1. Deactivate the Jetpack sharing module
2. Visit a restricted article as a member and confirm the "Gift this article" buttons render as expected

### Custom CTA settings

1. Edit the Content Gifting settings to use a different style, text, and URL
2. Visit the gifted article URL and confirm that the changes are applied and the button links to the URL

### Metering

Metering is already a form of gifting, so you can't gift metered content.

1. Enable metering
2. Confirm the "Gift this Article" button no longer renders

### Gifting limits

1. Disable metering again
2. Change the limit of gifting to 1
3. As a member, visit another gifted article
4. Click the "Gift this Article" button and confirm that the link is disabled

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

miguelpeixe committed:
feat: content gifting

miguelpeixe committed:
Merge branch 'trunk' into feat/content-gifting

miguelpeixe committed:
fix: overlay gate handling, metering and key data matching

miguelpeixe committed:
fix: styles and modal

miguelpeixe committed:
feat: settings

miguelpeixe committed:
Merge branch 'trunk' into feat/content-gifting

miguelpeixe committed:
fix: restore logic and move gate class

miguelpeixe committed:
fix: import and new restriction

thomasguillot committed:
feat(ui-icons): update gift

miguelpeixe committed:
Merge branch 'trunk' into feat/content-gifting

miguelpeixe committed:
fix: modal overlay management

miguelpeixe committed:
feat: ajax modal and jetpack share service

miguelpeixe committed:
fix: var name

miguelpeixe committed:
fix: notice margin

miguelpeixe committed:
fix: restore default value

miguelpeixe committed:
feat: settings

miguelpeixe committed:
fix: error message

miguelpeixe committed:
Merge branch 'trunk' into feat/content-gifting

miguelpeixe committed:
fix: only render gift button on singular pages

miguelpeixe committed:
fix: button margin

miguelpeixe committed:
chore: lint

miguelpeixe committed:
chore: lint

miguelpeixe committed:
feat: gift cta

miguelpeixe committed:
fix: padding

miguelpeixe committed:
fix: only render CTA if gift is being used

miguelpeixe committed:
feat: `post_has_restrictions` method

miguelpeixe committed:
fix: expiration and unit tests

miguelpeixe committed:
fix: icon only render

miguelpeixe committed:
fix: tweak limit messaging

thomasguillot committed:
feat(action-card): add togglePosition prop (leading | trailing)

thomasguillot committed:
Merge trunk into feat/content-gifting

thomasguillot committed:
feat: adjust icon-only style

thomasguillot committed:
feat: update inside jetpack; adjust copied label; add newspack-ui

thomasguillot committed:
fix: broken jetpack alignment

thomasguillot committed:
fix: typo

thomasguillot committed:
feat: fallback to newspack-ui colors instead of jetpack

thomasguillot committed:
feat(newspack-ui): update spacing var

thomasguillot committed:
feat: update banner style

thomasguillot committed:
feat: add banner style selector: light (default) or dark

miguelpeixe committed:
Merge branch 'trunk' into feat/content-gifting

miguelpeixe committed:
fix: automatically handle jp sharing service toggle

miguelpeixe committed:
chore: rename method

thomasguillot committed:
feat(content-gifting): add limit message to same paragraph

thomasguillot committed:
fix: class missing from checkout-button; web-preview wrong height

thomasguillot committed:
feat: add ability to customise cta label, add preview

thomasguillot commented:
@miguelpeixe I added a few things:

* Ability to edit the "Subscribe now" label
* Direct preview of the banner (including Newspack UI in wp-admin)
* Reorganised the settings with headings

miguelpeixe commented:
Thank you! It looks great and tested well. I've updated the PR screenshot to match the new UI.

thomasguillot committed:
feat(my-account): use newspack-ui in the sidebar (#4302)

matticbot committed:
chore: update translation files [skip ci]

laurelfulford committed:
fix: prevent spawning of Accessibility Statement pages (#4237)

matticbot committed:
chore: update translation files [skip ci]

dkoo committed:
fix: content gate metadata methods and params

semantic-release-bot committed:
chore(release): 6.24.2 [skip ci]

## [6.24.2](https://github.com/Automattic/newspack-plugin/compare/v6.24.1...v6.24.2) (2025-11-13)

### Bug Fixes

* content gate metadata methods and params ([d6344b2](https://github.com/Automattic/newspack-plugin/commit/d6344b29d1f83009c1ca9f680141906920c85351))
* content gate metadata methods and params [#4306](https://github.com/Automattic/newspack-plugin/issues/4306) from Automattic/hotfix/gate-post-metadata-nppm-2389 ([96615a1](https://github.com/Automattic/newspack-plugin/commit/96615a1368b569387cabc927f1c735c210351655))

dkoo committed:
fix(nrh): validation error in Donate block when no woocommerce (#4297)

semantic-release-bot committed:
chore(release): 6.24.3 [skip ci]

## [6.24.3](https://github.com/Automattic/newspack-plugin/compare/v6.24.2...v6.24.3) (2025-11-13)

### Bug Fixes

* **nrh:** validation error in Donate block when no woocommerce ([#4297](https://github.com/Automattic/newspack-plugin/issues/4297)) ([224fcea](https://github.com/Automattic/newspack-plugin/commit/224fcea40cee7b2453cc82e9121ce629d5745186))

miguelpeixe committed:
feat(ui): spinner tweaks (#4300)

miguelpeixe committed:
feat(cover-fees): support non-donation purchases (#4298)

matticbot committed:
chore: update translation files [skip ci]

miguelpeixe committed:
fix: class name

dkoo commented on src/content-gate/content-gifting.scss:
```diff
@@ -0,0 +1,125 @@
+}
+
+.sd-social-icon .sd-content ul li[class*="share-"].share-newspack-gift-article a.sd-button {
+	background: var(--newspack-theme-color-primary-against-white, var(--newspack-ui-color-primary-60, #{colors.$primary-600})) !important;
+	color: var(--newspack-theme-color-against-primary, var(--newspack-ui-color-neutral-0, #{colors.$neutral-000})) !important;
```
With the "icon only" style enabled in Jetpack settings, colors aren't quite matching on the front-end vs. what I see in the Jetpack settings admin page.

Jetpack settings:
<img width="481" height="110" alt="Screenshot 2025-11-13 at 1 54 47 PM" src="https://github.com/user-attachments/assets/12ccba3b-d35a-4097-a1b1-8676eccddbb7" />

Front-end:
<img width="272" height="79" alt="Screenshot 2025-11-13 at 1 54 42 PM" src="https://github.com/user-attachments/assets/0854b3d7-e428-434c-9491-2c831a45808c" />

dkoo replied:
Also, when I select "Official buttons" the "gift this article" button disappears from the preview:

<img width="407" height="199" alt="Screenshot 2025-11-13 at 1 58 04 PM" src="https://github.com/user-attachments/assets/e9953a90-6a92-4f1a-8bbc-870486af634d" />

I do still see it on the front-end, though:
<img width="285" height="61" alt="Screenshot 2025-11-13 at 1 58 26 PM" src="https://github.com/user-attachments/assets/1a58a3e7-bf9e-4964-97a4-0fbce5a2ba98" />

miguelpeixe replied:
This was a tricky one. Jetpack hardcodes the handling of the official button styles. In 9bce65018a7f2c7a6f72b89d2f6c1d1f88f80511 I trick Jetpack into thinking this is a Print button so it doesn't remove its contents.

This is where it's handling the preview: https://github.com/Automattic/jetpack/blob/817c26ec371251dead8428dd43573f955feb9528/projects/plugins/jetpack/modules/sharedaddy/admin-sharing.js#L163-L188

dkoo replied:
Gotcha, the hacky solution works!
<img width="529" height="201" alt="Screenshot 2025-11-14 at 9 40 08 AM" src="https://github.com/user-attachments/assets/4f06abff-6e1a-49dc-91ab-da4556b23dae" />

dkoo commented on src/wizards/audience/views/setup/content-gating.js:
```diff
@@ -76,6 +84,124 @@ export default withWizardScreen( () => {
+							<TextControl
+								label={ __( 'Subscribe button URL', 'newspack-plugin' ) }
+								help={ __(
+									'URL for the subscribe button in the banner. If not provided, the primary subscription tier product will be used with modal checkout.',
```
Non-blocking: can we link "primary subscription tier product" to the relevant wizard page?

miguelpeixe replied:
Good call! Added in 64f1284c3936d642af6f13dc2c0caf7638bd890c

<img width="1075" height="794" alt="image" src="https://github.com/user-attachments/assets/06e7dbd4-c5b6-4773-b85f-ddf41195f940" />

miguelpeixe committed:
Merge branch 'trunk' into feat/content-gifting

dkoo reviewed (changes requested):
@miguelpeixe This is working well and the UI looks good! I just ran into a few minor issues, outlined below and in inline comments:

1. When using "light" styles, the color of the CTA button doesn't match between the preview and the front-end. For reference, in my site the primary color is `#FF8200` (the orange color).

Preview:
<img width="1028" height="118" alt="Screenshot 2025-11-13 at 2 51 44 PM" src="https://github.com/user-attachments/assets/70a47f60-81e2-476c-9db8-f025783a0e42" />

Front-end:
<img width="1236" height="56" alt="Screenshot 2025-11-13 at 2 50 00 PM" src="https://github.com/user-attachments/assets/5e390900-1950-4456-afba-4eb23648b6af" />

3.  On the front-end, the banner is overlaid on top of the bottom-most edge of the viewport, which means it can obscure elements that are down there. I think we should try to add some bottom-padding to the `body` or a top-level container element equal to the height of the banner when it's rendered, so that it doesn't obscure anything.

<img width="1439" height="416" alt="Screenshot 2025-11-13 at 2 01 07 PM" src="https://github.com/user-attachments/assets/cfe504ac-3bdf-4906-a44b-aec03944ae86" />

4. It looks like I can use the same gift link to access an article multiple times in different anonymous browser sessions—so a single gift link once generated can be shared more or less infinitely. Is this intentional?

5. Maybe we should strip the `?content_key` URL param after it's been processed? That might also help control reuse of the same gift link a bit.

miguelpeixe committed:
fix: cta body offset

miguelpeixe committed:
fix: enqueue theme colors css in the admin

miguelpeixe committed:
feat: replace query param with cookie on load

miguelpeixe committed:
fix: official preview

miguelpeixe committed:
feat: link to primary subscription product

miguelpeixe commented:
Thanks for the review!

> When using "light" styles, the color of the CTA button doesn't match between the preview and the front-end. For reference, in my site the primary color is #FF8200 (the orange color).

We were missing the theme CSS variables to match the Newspack UI experience in the admin area. Added in da2fbd5bb7ae8677ef61f451f745db61ec7a0764

> With the "icon only" style enabled in Jetpack settings, colors aren't quite matching on the front-end vs. what I see in the Jetpack settings admin page.

With the change mentioned above, they should now match. They'll match with the odd color, though... Maybe something about [this calculation](https://github.com/Automattic/newspack-theme/blob/b2dbc607fdcbb70952a186b71ac6b7b81d4b994d/newspack-theme/inc/color-patterns.php#L28) from theme?

> On the front-end, the banner is overlaid on top of the bottom-most edge of the viewport, which means it can obscure elements that are down there. I think we should try to add some bottom-padding to the body or a top-level container element equal to the height of the banner when it's rendered, so that it doesn't obscure anything.

Good call! Added a padding with dynamic offset detection to account for mobile and dynamic CTA content in 9317fc3ca7c7d0a202bd53b3941f95463233ea46

> It looks like I can use the same gift link to access an article multiple times in different anonymous browser sessions—so a single gift link once generated can be shared more or less infinitely. Is this intentional?
> 
> Maybe we should strip the ?content_key URL param after it's been processed? That might also help control reuse of the same gift link a bit.

Yes, anyone can use the link until the 24-hour expiration. I like the idea of stripping the query param. I implemented that with a cookie replacement, so the recipient can still visit the article for the granted time. 97f65186901d6b3c9c20c7c2ec0118732faadbab

The cookie starts with `wp_`, so batcache skips it.

dkoo reviewed (changes requested):
Thanks for the changes, @miguelpeixe! This is almost ready now, the only thing is I'm still seeing some color discrepancies between preview vs. front-end:

For the CTA in the banner:

| preview | front-end |
| ---- | ---- |
| <img width="1024" height="118" alt="Screenshot 2025-11-14 at 9 33 10 AM" src="https://github.com/user-attachments/assets/8d96c819-c9f9-4fc0-9380-f3ef30a02258" /> | <img width="1428" height="71" alt="Screenshot 2025-11-14 at 10 05 04 AM" src="https://github.com/user-attachments/assets/061d78a3-2646-492f-86f1-4f9a5eca27fc" /> |

With Jetpack sharing + icon-only styles:

| preview | front-end |
| ---- | ---- |
| <img width="457" height="222" alt="Screenshot 2025-11-14 at 10 08 07 AM" src="https://github.com/user-attachments/assets/978769e5-9bd8-4a92-9670-9fe1b2c8be7a" /> | <img width="256" height="85" alt="Screenshot 2025-11-14 at 9 43 10 AM" src="https://github.com/user-attachments/assets/fd873328-2636-4e9e-8112-ba9a91646485" /> |

Other than that, everything works as described!

miguelpeixe committed:
fix: missing theme variables

miguelpeixe commented:
Thanks, @dkoo! I missed a couple of theme variables. Added in 13783c57e624dcc563f5b97d018fd95980c6bd4e
